rclpy/rclpy/impl/logger_rcutils.py

```diff
@@ -0,0 +1,275 @@
+
+
+def _find_caller(frame):
+    """Get the first calling frame that is outside of rclpy."""
+    global _internal_callers
+    if _internal_callers is None:
+        # Populate the list of known filenames from which logging methods can be called.
+        # This has to be done from within a function to avoid cyclic module imports.
+        import rclpy.logging
+        _internal_callers = [__file__, rclpy.logging.__file__]
+
+    frame_filename = inspect.getabsfile(frame)
+    while any(f in frame_filename for f in _internal_callers):
+        frame = frame.f_back
+        frame_filename = inspect.getabsfile(frame)
+    return frame
+
+
+class CallerId:
+
+    def __init__(self, frame=None):
+        if not frame:
+            frame = _find_caller(inspect.currentframe())
+        self.function_name = frame.f_code.co_name
+        self.file_name = inspect.getabsfile(frame)
+        self.line_number = frame.f_lineno
+        self.last_index = frame.f_lasti  # To distinguish between two callers on the same line
+
+
+class LoggingFilter:
+    """Base class for logging filters."""
+
+    """
+    Parameters of a filter and their default value, if appropriate.
+
+    A default value of None makes a parameter required.
+    """
+    params = {}
+
+    """
+    Initialize the context of a logging call, e.g. declare variables needed for
+    determining the log condition and add them to the context.
+    """
+    @classmethod
+    def initialize_context(cls, context, **kwargs):
+        # Store all parameters in the context so we can check that users never try to change them.
+        for param in cls.params:
+            context[param] = kwargs.get(param, cls.params[param])
+            if context[param] is None:
+                raise RuntimeError(
+                    'Required parameter "{0}" was not specified for logging filter "{1}"'
+                    .format(param, cls.__name__))
+
+    """
+    Decide if it's appropriate to log given a context, and update the context accordingly.
+    """
+    @staticmethod
+    def log_condition(context):
+        return True
+
+
+class Once(LoggingFilter):
+    """Ignore all log calls except the first one."""
+
+    params = {
+        'once': None,
+    }
+
+    @classmethod
+    def initialize_context(cls, context, **kwargs):
+        super(cls, cls).initialize_context(context, **kwargs)
+        context['has_been_logged_once'] = False
+
+    @staticmethod
+    def log_condition(context):
+        logging_condition = False
+        if not context['has_been_logged_once']:
+            logging_condition = True
+            context['has_been_logged_once'] = True
+        return logging_condition
+
+
+class Throttle(LoggingFilter):
+    """Ignore log calls if the last call is not longer ago than the specified duration."""
+
+    params = {
+        'throttle_duration_sec': None,
+        'throttle_time_source_type': 'RCUTILS_STEADY_TIME',
```

Comment:
the logic is in python but (in theory) it's based on rcutils time underneath. at the moment it's using internal python time but I understood that one day we'd expose rcutils time concepts in rclpy. Even though it's not really using rcutils steady time at the moment, I didn't want to put an honest default because we might not keep support for that. I added an explanation for users changing the option to not expect it to work with other values in 95dbecf9c076c5ac7768c332662b615d02ee5eda